// @sentry/cloudflare/build/cjs/durableobject.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const core = require('@sentry/core');
const async = require('./async.js');
const instrument = require('./instrument.js');
const options = require('./options.js');
const request = require('./request.js');
const sdk = require('./sdk.js');
const copyExecutionContext = require('./utils/copyExecutionContext.js');

/* eslint-disable @typescript-eslint/unbound-method */

function wrapMethodWithSentry(
  wrapperOptions,
  handler,
  callback,
  noMark,
) {
  if (instrument.isInstrumented(handler)) {
    return handler;
  }

  if (!noMark) {
    instrument.markAsInstrumented(handler);
  }

  return new Proxy(handler, {
    apply(target, thisArg, args) {
      const currentClient = core.getClient();
      // if a client is already set, use withScope, otherwise use withIsolationScope
      const sentryWithScope = currentClient ? core.withScope : core.withIsolationScope;

      const wrappedFunction = (scope) => {
        // In certain situations, the passed context can become undefined.
        // For example, for Astro while prerendering pages at build time.
        // see: https://github.com/getsentry/sentry-javascript/issues/13217
        const context = wrapperOptions.context ;

        const waitUntil = context?.waitUntil?.bind?.(context);

        const currentClient = scope.getClient();
        if (!currentClient) {
          const client = sdk.init({ ...wrapperOptions.options, ctx: context });
          scope.setClient(client);
        }

        if (!wrapperOptions.spanName) {
          try {
            if (callback) {
              callback(...args);
            }
            const result = Reflect.apply(target, thisArg, args);

            if (core.isThenable(result)) {
              return result.then(
                (res) => {
                  waitUntil?.(core.flush(2000));
                  return res;
                },
                (e) => {
                  core.captureException(e, {
                    mechanism: {
                      type: 'auto.faas.cloudflare.durable_object',
                      handled: false,
                    },
                  });
                  waitUntil?.(core.flush(2000));
                  throw e;
                },
              );
            } else {
              waitUntil?.(core.flush(2000));
              return result;
            }
          } catch (e) {
            core.captureException(e, {
              mechanism: {
                type: 'auto.faas.cloudflare.durable_object',
                handled: false,
              },
            });
            waitUntil?.(core.flush(2000));
            throw e;
          }
        }

        const attributes = wrapperOptions.spanOp
          ? {
              [core.SEMANTIC_ATTRIBUTE_SENTRY_OP]: wrapperOptions.spanOp,
              [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.faas.cloudflare.durable_object',
            }
          : {};

        return core.startSpan({ name: wrapperOptions.spanName, attributes }, () => {
          try {
            const result = Reflect.apply(target, thisArg, args);

            if (core.isThenable(result)) {
              return result.then(
                (res) => {
                  waitUntil?.(core.flush(2000));
                  return res;
                },
                (e) => {
                  core.captureException(e, {
                    mechanism: {
                      type: 'auto.faas.cloudflare.durable_object',
                      handled: false,
                    },
                  });
                  waitUntil?.(core.flush(2000));
                  throw e;
                },
              );
            } else {
              waitUntil?.(core.flush(2000));
              return result;
            }
          } catch (e) {
            core.captureException(e, {
              mechanism: {
                type: 'auto.faas.cloudflare.durable_object',
                handled: false,
              },
            });
            waitUntil?.(core.flush(2000));
            throw e;
          }
        });
      };

      return sentryWithScope(wrappedFunction);
    },
  });
}

/**
 * Instruments a Durable Object class to capture errors and performance data.
 *
 * Instruments the following methods:
 * - fetch
 * - alarm
 * - webSocketMessage
 * - webSocketClose
 * - webSocketError
 *
 * as well as any other public RPC methods on the Durable Object instance.
 *
 * @param optionsCallback Function that returns the options for the SDK initialization.
 * @param DurableObjectClass The Durable Object class to instrument.
 * @returns The instrumented Durable Object class.
 *
 * @example
 * ```ts
 * class MyDurableObjectBase extends DurableObject {
 *   constructor(ctx: DurableObjectState, env: Env) {
 *     super(ctx, env);
 *   }
 * }
 *
 * export const MyDurableObject = instrumentDurableObjectWithSentry(
 *   env => ({
 *     dsn: env.SENTRY_DSN,
 *     tracesSampleRate: 1.0,
 *   }),
 *   MyDurableObjectBase,
 * );
 * ```
 */
function instrumentDurableObjectWithSentry

(optionsCallback, DurableObjectClass) {
  return new Proxy(DurableObjectClass, {
    construct(target, [ctx, env]) {
      async.setAsyncLocalStorageAsyncContextStrategy();
      const context = copyExecutionContext.copyExecutionContext(ctx);

      const options$1 = options.getFinalOptions(optionsCallback(env), env);

      const obj = new target(context, env);

      // These are the methods that are available on a Durable Object
      // ref: https://developers.cloudflare.com/durable-objects/api/base/
      // obj.alarm
      // obj.fetch
      // obj.webSocketError
      // obj.webSocketClose
      // obj.webSocketMessage

      // Any other public methods on the Durable Object instance are RPC calls.

      if (obj.fetch && typeof obj.fetch === 'function' && !instrument.isInstrumented(obj.fetch)) {
        obj.fetch = new Proxy(obj.fetch, {
          apply(target, thisArg, args) {
            return request.wrapRequestHandler({ options: options$1, request: args[0], context }, () =>
              Reflect.apply(target, thisArg, args),
            );
          },
        });

        instrument.markAsInstrumented(obj.fetch);
      }

      if (obj.alarm && typeof obj.alarm === 'function') {
        obj.alarm = wrapMethodWithSentry({ options: options$1, context, spanName: 'alarm' }, obj.alarm);
      }

      if (obj.webSocketMessage && typeof obj.webSocketMessage === 'function') {
        obj.webSocketMessage = wrapMethodWithSentry(
          { options: options$1, context, spanName: 'webSocketMessage' },
          obj.webSocketMessage,
        );
      }

      if (obj.webSocketClose && typeof obj.webSocketClose === 'function') {
        obj.webSocketClose = wrapMethodWithSentry({ options: options$1, context, spanName: 'webSocketClose' }, obj.webSocketClose);
      }

      if (obj.webSocketError && typeof obj.webSocketError === 'function') {
        obj.webSocketError = wrapMethodWithSentry(
          { options: options$1, context, spanName: 'webSocketError' },
          obj.webSocketError,
          (_, error) =>
            core.captureException(error, {
              mechanism: {
                type: 'auto.faas.cloudflare.durable_object_websocket',
                handled: false,
              },
            }),
        );
      }

      for (const method of Object.getOwnPropertyNames(obj)) {
        if (
          method === 'fetch' ||
          method === 'alarm' ||
          method === 'webSocketError' ||
          method === 'webSocketClose' ||
          method === 'webSocketMessage'
        ) {
          continue;
        }

        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        const value = (obj )[method] ;
        if (typeof value === 'function') {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
          (obj )[method] = wrapMethodWithSentry(
            { options: options$1, context, spanName: method, spanOp: 'rpc' },
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            value ,
          );
        }
      }

      // Store context and options on the instance for prototype methods to access
      Object.defineProperty(obj, '__SENTRY_CONTEXT__', {
        value: context,
        enumerable: false,
        writable: false,
        configurable: false,
      });

      Object.defineProperty(obj, '__SENTRY_OPTIONS__', {
        value: options$1,
        enumerable: false,
        writable: false,
        configurable: false,
      });

      if (options$1?.instrumentPrototypeMethods) {
        instrumentPrototype(target, options$1.instrumentPrototypeMethods);
      }

      return obj;
    },
  });
}

function instrumentPrototype(target, methodsToInstrument) {
  const proto = target.prototype;

  // Get all methods from the prototype chain
  const methodNames = new Set();
  let current = proto;

  while (current && current !== Object.prototype) {
    Object.getOwnPropertyNames(current).forEach(name => {
      if (name !== 'constructor' && typeof (current )[name] === 'function') {
        methodNames.add(name);
      }
    });
    current = Object.getPrototypeOf(current);
  }

  // Create a set for efficient lookups when methodsToInstrument is an array
  const methodsToInstrumentSet = Array.isArray(methodsToInstrument) ? new Set(methodsToInstrument) : null;

  // Instrument each method on the prototype
  methodNames.forEach(methodName => {
    const originalMethod = (proto )[methodName];

    if (!originalMethod || instrument.isInstrumented(originalMethod)) {
      return;
    }

    // If methodsToInstrument is an array, only instrument methods in that set
    if (methodsToInstrumentSet && !methodsToInstrumentSet.has(methodName)) {
      return;
    }

    // Create a wrapper that gets context/options from the instance at runtime
    const wrappedMethod = function ( ...args) {
      const thisWithSentry = this

;
      const instanceContext = thisWithSentry.__SENTRY_CONTEXT__;
      const instanceOptions = thisWithSentry.__SENTRY_OPTIONS__;

      if (!instanceOptions) {
        // Fallback to original method if no Sentry data found
        return (originalMethod ).apply(this, args);
      }

      // Use the existing wrapper but with instance-specific context/options
      const wrapper = wrapMethodWithSentry(
        {
          options: instanceOptions,
          context: instanceContext,
          spanName: methodName,
          spanOp: 'rpc',
        },
        originalMethod ,
        undefined,
        true, // noMark = true since we'll mark the prototype method
      );

      return (wrapper ).apply(this, args);
    };

    instrument.markAsInstrumented(wrappedMethod);

    // Replace the prototype method
    Object.defineProperty(proto, methodName, {
      value: wrappedMethod,
      enumerable: false,
      writable: true,
      configurable: true,
    });
  });
}

exports.instrumentDurableObjectWithSentry = instrumentDurableObjectWithSentry;
