// @sentry/cloudflare/build/cjs/integrations/hono.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const core = require('@sentry/core');
const debugBuild = require('../debug-build.js');

const INTEGRATION_NAME = 'Hono';

/** Only exported for internal use */
function getHonoIntegration() {
  return core.getClient()?.getIntegrationByName(INTEGRATION_NAME);
}

function isHonoError(err) {
  if (err instanceof Error) {
    return true;
  }
  return typeof err === 'object' && err !== null && 'status' in (err );
}

const _honoIntegration = ((options = {}) => {
  return {
    name: INTEGRATION_NAME,
    handleHonoException(err) {
      const shouldHandleError = options.shouldHandleError || defaultShouldHandleError;

      if (!isHonoError(err)) {
        debugBuild.DEBUG_BUILD && core.debug.log("[Hono] Won't capture exception in `onError` because it's not a Hono error.", err);
        return;
      }

      if (shouldHandleError(err)) {
        core.captureException(err, { mechanism: { handled: false, type: 'auto.faas.hono.error_handler' } });
      } else {
        debugBuild.DEBUG_BUILD && core.debug.log('[Hono] Not capturing exception because `shouldHandleError` returned `false`.', err);
      }
    },
  };
}) ;

/**
 * Automatically captures exceptions caught with the `onError` handler in Hono.
 *
 * The integration is enabled by default.
 *
 * @example
 * integrations: [
 *   honoIntegration({
 *     shouldHandleError: (err) => true; // always capture exceptions in onError
 *   })
 * ]
 */
const honoIntegration = core.defineIntegration(_honoIntegration);

/**
 * Default function to determine if an error should be sent to Sentry
 *
 * 3xx and 4xx errors are not sent by default.
 */
function defaultShouldHandleError(error) {
  const statusCode = error?.status;
  // 3xx and 4xx errors are not sent by default.
  return statusCode ? statusCode >= 500 || statusCode <= 299 : true;
}

exports.getHonoIntegration = getHonoIntegration;
exports.honoIntegration = honoIntegration;
